File: server/services/gaanapyService.js
const GAANAPY_BASE_URL = process.env.GAANAPY_BASE_URL || 'http://127.0.0.1:8000';

function buildUrl(path, params = {}) {
  const url = new URL(path, GAANAPY_BASE_URL);
  for (const [key, value] of Object.entries(params)) {
    if (value !== undefined && value !== null && value !== '') {
      url.searchParams.set(key, String(value));
    }
  }
  return url;
}

function unwrapResults(body) {
  if (Array.isArray(body)) return body;
  if (Array.isArray(body?.results)) return body.results;
  if (Array.isArray(body?.tracks)) return body.tracks;
  if (Array.isArray(body?.albums)) return body.albums;
  if (Array.isArray(body?.artists)) return body.artists;
  if (Array.isArray(body?.entities)) return body.entities;
  return [];
}

function normalizeTrack(track = {}) {
  const streamUrls = track.stream_urls?.urls || {};
  const imageUrls = track.images?.urls || {};

  return {
    id: String(track.track_id || track.seokey || track.id || track.song_id || Date.now()),
    seokey: track.seokey || String(track.track_id || track.id || ''),
    trackId: String(track.track_id || track.id || ''),
    title: track.title || track.track_title || track.name || 'Unknown Track',
    artistName: track.artists || track.artistName || track.artist_name || '',
    albumName: track.album || track.albumName || track.album_title || '',
    duration: Number(track.duration || 0),
    thumbnail: imageUrls.large_artwork || imageUrls.medium_artwork || track.artist_image || track.thumbnail || track.image || '',
    streamUrl: streamUrls.very_high_quality || streamUrls.high_quality || streamUrls.medium_quality || streamUrls.low_quality || track.audio || track.streamUrl || '',
    audioDownloadUrl: track.audioDownloadUrl || track.audiodownload || '',
    shareUrl: track.song_url || track.shareUrl || track.shorturl || '',
    licenseUrl: track.licenseUrl || '',
    source: 'gaanapy',
    albumUrl: track.album_url || track.albumUrl || '',
    songUrl: track.song_url || track.songUrl || '',
    popularity: track.popularity || '',
    language: track.language || '',
  };
}

async function requestJson(path, params = {}) {
  const response = await fetch(buildUrl(path, params));
  if (!response.ok) {
    throw new Error(`GaanaPy request failed with ${response.status}`);
  }

  const body = await response.json();
  return body;
}

async function fetchTracks(path, params = {}) {
  const body = await requestJson(path, params);
  const results = unwrapResults(body);
  return results.map(normalizeTrack).filter((track) => Boolean(track.streamUrl));
}

export async function searchTracks(query, limit = 24) {
  const trimmed = query.trim();
  if (!trimmed) return [];

  return fetchTracks('/songs/search/', {
    query: trimmed,
    limit: String(limit),
  });
}

export async function getTrackById(trackId) {
  if (!trackId) return null;

  try {
    const body = await requestJson('/songs/info/', { seokey: trackId });
    const results = unwrapResults(body);
    if (results.length > 0) return normalizeTrack(results[0]);

    if (body && typeof body === 'object' && !Array.isArray(body) && body.seokey) {
      return normalizeTrack(body);
    }
  } catch (error) {
    return null;
  }

  return null;
}

export async function getTrendingTracks(limit = 12, language = 'Hindi') {
  return fetchTracks('/trending', {
    language,
    limit: String(limit),
  });
}

export async function getNewReleaseTracks(limit = 12, language = 'Hindi') {
  const body = await requestJson('/newreleases', {
    language,
    limit: String(limit),
  });

  const results = unwrapResults(body);
  return results.map(normalizeTrack).filter((track) => Boolean(track.streamUrl));
}

export async function getCharts(limit = 12) {
  const body = await requestJson('/charts', { limit: String(limit) });
  const results = unwrapResults(body);
  return results.map(normalizeTrack).filter((track) => Boolean(track.streamUrl));
}

File: server/app/gaanapy_service.py
from typing import Optional
from server.GaanaPy.api.gaanapy import GaanaPy
from .models import Track

gaanapy = GaanaPy()


def _extract_stream_url(track: dict) -> str:
    stream_urls = track.get("stream_urls", {})
    urls = stream_urls.get("urls", {}) if isinstance(stream_urls, dict) else {}
    if isinstance(urls, dict):
        for key in ("very_high_quality", "high_quality", "medium_quality", "low_quality"):
            url = urls.get(key)
            if url:
                return url
    return track.get("stream_url", "") or ""


def _normalize(track: dict) -> Optional[Track]:
    try:
        duration = int(track.get("duration", 0))
    except (ValueError, TypeError):
        duration = 0

    images = track.get("images", {})
    urls = images.get("urls", {}) if isinstance(images, dict) else {}
    thumbnail = (
        urls.get("medium_artwork")
        or urls.get("large_artwork")
        or urls.get("small_artwork")
        or ""
    )

    artist = track.get("artists", track.get("artist", "Unknown Artist"))
    title = track.get("title", "Unknown Track")
    track_id = track.get("seokey") or track.get("track_id") or str(hash(title))

    return Track(
        id=track_id,
        title=title,
        artist=artist if isinstance(artist, str) else ", ".join(artist),
        album=track.get("album", ""),
        duration=float(duration),
        thumbnail=thumbnail,
        stream_url=_extract_stream_url(track),
        source="gaanapy",
    )


def _normalize_list(results: object) -> list[Track]:
    if not isinstance(results, list):
        return []
    tracks = []
    for item in results:
        normalized = _normalize(item)
        if normalized:
            tracks.append(normalized)
    return tracks


async def search_tracks(query: str, limit: int = 12) -> list[Track]:
    results = await gaanapy.search_songs(query, limit)
    return _normalize_list(results)


async def get_track(track_id: str) -> Optional[Track]:
    results = await gaanapy.get_track_info([track_id])
    if isinstance(results, list) and results:
        return _normalize(results[0])
    return None


async def get_trending(language: str = "Hindi", limit: int = 12) -> list[Track]:
    results = await gaanapy.get_trending(language, limit)
    return _normalize_list(results)


async def get_new_releases(language: str = "Hindi", limit: int = 12) -> list[Track]:
    results = await gaanapy.get_new_releases(language, limit)
    tracks = results.get("tracks", []) if isinstance(results, dict) else results
    return _normalize_list(tracks)


async def get_charts(limit: int = 12) -> list[Track]:
    results = await gaanapy.get_charts(limit)
    return _normalize_list(results)


async def close_gaanapy():
    await gaanapy.close()
